### PR TITLE
Fix missing scalar issue when evaluating subqueries inside table functions

### LINKED ISSUE #56031: Scalar doesn't exist for `format` function
```sql
SELECT *
FROM format(TSVRaw, (
    SELECT
        '123',
        '456'
))
FORMAT TSVRaw

Query id: c5e0b9f8-d055-4b5d-8e14-b695f8096ab0


0 rows in set. Elapsed: 0.002 sec.

Received exception:
Code: 36. DB::Exception: Scalar `2451974040954057853_15236037560691200747` doesn't exist (internal bug): While processing __getScalar('2451974040954057853_15236037560691200747') AS constant_expression. (BAD_ARGUMENTS)
```

### PR DESCRIPTION
**Pull request by amosbird**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix missing scalar issue when evaluating subqueries inside table functions. This fixes #56031.

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 829fc61e7736582585ccbe99dde63af023c1582e with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56057/829fc61e7736582585ccbe99dde63af023c1582e/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6657893611">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/56057/829fc61e7736582585ccbe99dde63af023c1582e/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by Algunenano** [comment]
Failures:
* `Bugfix validate check — Changed tests don't reproduce the bug`. It complains because the previous release passes the test, as you are only checking the ErrorCode and that doesn't change. You can either tweak a bit the test or change the PR to an improvement.
* `Stateless tests (release, wide parts enabled) ` -> `02263_lazy_mark_load` -> https://github.com/ClickHouse/ClickHouse/issues/55973
* `Upgrade check (msan)` -> A LOGICAL_ERROR due to using distributed tables and allow_experimental_partial_result together (`23.9.2.56`). The feature had to be removed because of these kind of problems in master.
* `Upgrade check (tsan)` -> Cannot start the new CH server. The problem is that the previous one never finished shutting down and we didn't kill it.

**Review by Algunenano — COMMENTED** [review]
Do you mind adapting the test so that it doesn't pass with the previous release or changing the PR to an improvement , please?

**Comment by amosbird** [comment]
Let's see if this PR can also fix the error code.

File: src/Interpreters/Context.cpp [hunk 1]
```diff
@@ -1400,9 +1400,7 @@ const Block & Context::getScalar(const String & name) const
     auto it = scalars.find(name);
     if (scalars.end() == it)
     {
-        // This should be a logical error, but it fails the sql_fuzz test too
-        // often, so 'bad arguments' for now.
-        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Scalar {} doesn't exist (internal bug)", backQuoteIfNeed(name));
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Scalar {} doesn't exist (internal bug)", backQuoteIfNeed(name));
```
Comment (Algunenano) [inline src/Interpreters/Context.cpp:1403]: Sadly this is not ok: `SELECT __getScalar('')`

**Review by Algunenano — APPROVED** [review]

**Pull request merged**